### PR TITLE
Support listing keys in custom metadata map via C/C++ API

### PR DESCRIPTION
**Pull request by hariharans29**
Missing one API from request in https://github.com/microsoft/onnxruntime/issues/2746

### CODE REVIEW

**Comment by hariharans29** [comment]
CC: @ankane

**Comment by ankane** [comment]
Awesome, thanks @hariharans29!

File: onnxruntime/core/session/onnxruntime_c_api.cc [hunk 1]
```diff
@@ -755,6 +755,32 @@ ORT_API_STATUS_IMPL(OrtApis::ModelMetadataLookupCustomMetadataMap,
   API_IMPL_END
 }
 
+ORT_API_STATUS_IMPL(OrtApis::ModelMetadataGetCustomMetadataMapKeys,
+                    _In_ const OrtModelMetadata* model_metadata,
+                    _Inout_ OrtAllocator* allocator, _Outptr_ char*** keys, _Out_ int64_t* num_keys) {
+  API_IMPL_BEGIN
+  auto custom_metadata_map =
```
Comment (pranavsharma) [inline onnxruntime/core/session/onnxruntime_c_api.cc:762]: const auto*

File: onnxruntime/core/session/onnxruntime_c_api.cc [hunk 2]
```diff
@@ -755,6 +755,32 @@ ORT_API_STATUS_IMPL(OrtApis::ModelMetadataLookupCustomMetadataMap,
...
+      reinterpret_cast<const ::onnxruntime::ModelMetadata*>(model_metadata)->custom_metadata_map;
+
+  auto count = custom_metadata_map.size();
+  if (count == 0) {
+    *keys = nullptr;
+  } else {
+    *keys = reinterpret_cast<char**>(allocator->Alloc(allocator, count * sizeof(char*)));
```
Comment (pranavsharma) [inline onnxruntime/core/session/onnxruntime_c_api.cc:769]: Use SafeInt when calling Alloc. There are plenty of such examples in the codebase. Also see https://github.com/microsoft/onnxruntime/blob/master/docs/Coding_Conventions_and_Standards.md.

**Comment by Craigacp** [comment]
Is this going to land before the next release? I'm currently writing the Java support for metadata, and being able to iterate the keys would be very useful.

**Comment by hariharans29** [comment]
> Is this going to land before the next release? I'm currently writing the Java support for metadata, and being able to iterate the keys would be very useful.

It should go in before the next release. It is just pending a review.

File: onnxruntime/core/session/onnxruntime_c_api.cc (hunk 2 above)
Comment (hariharans29) [reply onnxruntime/core/session/onnxruntime_c_api.cc]: Done, thanks!

**Comment by Craigacp** [comment]
Ok. I had assumed that the C API `ORT_API_VERSION` number was only going to be bumped with major version changes as the C API refroze. Is that not the case?

**Comment by pranavsharma** [comment]
> Ok. I had assumed that the C API `ORT_API_VERSION` number was only going to be bumped with major version changes as the C API refroze. Is that not the case?

This is really the version of the struct that is holding the C API function ptrs. The semantic versioning is still intact.

**Pull request merged**